### ISSUE TITLE
Make use of sander API in energy calculation

### ISSUE THREAD

**Issue by drroe** [Amber-MD/cpptraj#248]
This is actually an ongoing issue; there is already an option to compile with the sander API, although in its current state it does not work. I've got a branch now where things seems to be working ok. The ultimate goal would be having something that could rival or maybe even replace imin=5 post-processing.

**Comment by hainm** [comment]
imin=1 is useful too. I need to minimize hundreds of structures and need to generate rst7 file for each.
It would be nice if just need prmtop file, cpptraj reads a bunch of pdb files, run imin=1 and write to new rst7 files, or single trajectory.

PS: imin=1 is in @swails todo-list.

**Comment by swails** [comment]
> PS: imin=1 is in @swails todo-list.

It is?

**Comment by swails** [comment]
> imin=1 is useful too. I need to minimize hundreds of structures and need to generate rst7 file for each.
> It would be nice if just need prmtop file, cpptraj reads a bunch of pdb files, run imin=1 and write to new rst7 files, or single trajectory.

This is actually imin=5.  imin=1 is a single structure only.  5 means minimize every snapshot for maxcyc number of steps.

**Comment by hainm** [comment]
> 5 means minimize every snapshot for maxcyc number of steps.

oh, I should read manual more carefully.

**Comment by hainm** [comment]
> It is?

uhm, I mis-remembered this issue https://github.com/ParmEd/ParmEd/issues/243

**Comment by drroe** [comment]
Currently being implemented via #249.

Everything seems to work well, I've done some testing on solvated and non-solvated (GB) trajectories and everything matches up to `imin=5` processing. This even works in parallel!

However, PB tends to "bomb" all over the place. That's not the fault of the API; it crashes out at similar points when using `imin=5` as well. Unfortunately there is no way to gracefully recover from these since the aborts are hard-coded into `pbsa`.

Once this is merged in I will have to do some testing on the AmberTools side and decide if this should be on by default or not.

**Comment by drroe** [comment]
Note to self: this is going to need more error checking to make sure parameters are present and the file is valid. Maybe add an option to write a temporary parm if no file available.

**Comment by drroe** [comment]
Turns out I can't just use pubfft from libsander - it's different enough from the pubfft in cpptraj (which itself is from pmemd) that even though `esander` works all other FFT functionality in cpptraj causes a segfault. To get around this I just changed the function names in cpptraj's pubfft so there will be no clashes and things seem to work.

**Comment by swails** [comment]
I *suppose* that's easier than completely swapping out a separate FFT implementation :smiley:

**Comment by drroe** [comment]
@swails @hainm I've got a bit of a dilemma here and I'd be interested to hear your takes on it.

I'm currently trying to decide how cpptraj should use the sander API when being built inside AmberTools. My initial thought was to just enable it by default since it's part of AmberTools. This is fine as long as the serial build is created first. However as we know that's not always the case - some people jump straight to the parallel build, and they should be able to do that. But since libsander.so is not built in parallel this would cause the cpptraj build to fail.

Another alternative is to have a separate build target, something like `cpptraj.sander`, that gets built if `libsander.so` exists and warns otherwise. I don't really like this because the *only* difference then between `cpptraj` and `cpptraj.sander` is the `esander` action, which seems silly to me.

The third option I've come up with is to treat the sander API like we do NetCDF and have it be built as part of configure, so it's always there. That will add yet more runtime to configure (at least the first time around) but may be the best option in the long run. What do you guys think?

**Comment by hainm** [comment]
> Another alternative is to have a separate build target, something like cpptraj.sander, that gets built if libsander.so exists and warns otherwise. I don't really like this because the only difference then between cpptraj and cpptraj.sander is the esander action, which seems silly to me.

I am also -1 for this. cpptraj already has many extensions (.so, .OMP, .MPI). (this is one of reasons why I decided to write my own parallel implementation in pytraj to avoid the complexity compiling different cpptraj version). 

> The third option I've come up with is to treat the sander API like we do NetCDF and have it be built as part of configure, so it's always there. That will add yet more runtime to configure (at least the first time around) but may be the best option in the long run. What do you guys think?

I am +1 for this option. pysander is used in phenix, pytraj, and now cpptraj. We are collaborating with Sagar's lab to use AMBER FF for protein design, and we use pysander to energy evaluation. So it's great benefit to have libsander.so installed as default.

> That will add yet more runtime to configure (at least the first time around) but may be the best option in the long run.

**more** == **much more**. But I think it's fine. If we (you or Jason) implement it, should we show the progress of compiling too?

I just check `-mpi` option, it requires to build serial version first.

```bash
    Most common flags:
      -mpi           Use MPI for parallelization. [Note: you must first
                     build the serial version before using this flag.]
```

**Comment by drroe** [comment]
Oh, I hadn't realized `serial` was required first. That actually takes care of the problem itself.

The build sander API option is attractive, but I think far too much work to implement and test before the code deadline. So I think I will go with option one for now.

**Comment by swails** [comment]
That's not true for everything, but a lot of parallel programs now depend on serial programs as well.  For instance, MMPBSA.py.MPI requires serial executables.  But the *installs* themselves won't die if serial version is not compiled first.

What about maybe making the sander API optional through the use of `dlopen` as a runtime dependency?

**Comment by hainm** [comment]
> What about maybe making the sander API optional through the use of dlopen as a runtime dependency?

:100: this also make pytraj's life easier.

**Comment by swails** [comment]
Would pytraj not continue to use pysander?

**Comment by hainm** [comment]
> Oh, I hadn't realized serial was required first. That actually takes care of the problem itself.

Not actuall, AMBER does not have strict checking if serial version are built first. I myself got the headache before (and users asked about this issue in maillist too)

**Comment by hainm** [comment]
> Would pytraj not continue to use pysander?

pytraj always use pysander.

Clarify: I meant `dlopen` in general case. (in another issue topic).

**Comment by hainm** [comment]
> pytraj always use pysander.

Aslo, it's easier to write parallel code with pysander (vs cpptraj  MPI + sander).

**Comment by drroe** [comment]
> Not actuall, AMBER does not have strict checking if serial version are built first. 

In the case of libsander it would, because libsander isn't built in parallel and the cpptraj build rule I'm testing requires it. So it will fail, and can fail with a nice error message.

**Comment by drroe** [comment]
> What about maybe making the sander API optional through the use of dlopen as a runtime dependency?

I don't have a lot of experience with this, but I would think from a user standpoint it would be pretty annoying to have a compile seemingly go well and then have cpptraj freak out during runtime because the right library isn't there. Unless I'm not correctly understanding how this would work.

**Comment by hainm** [comment]
> I don't have a lot of experience with this, but I would think from a user standpoint it would be pretty annoying to have a compile seemingly go well and then have cpptraj freak out during runtime because the right library isn't there.

This is how python works :D  

For example, pytraj compilation is fine without `pysander`.

But if user calls `pytraj.energy_decomposition`, they need to have pysander installed.

**Comment by swails** [comment]
> I don't have a lot of experience with this, but I would think from a user standpoint it would be pretty annoying to have a compile seemingly go well and then have cpptraj freak out during runtime because the right library isn't there. Unless I'm not correctly understanding how this would work.

It would only freak out if you tried *using* those symbols.  So you could have it print a message like "sander not available -- compile in serial".  And once they compiled it, it would start working (no need to recompile cpptraj).

`dlopen` really turns it into a runtime dependency, not a compile-time.

**Comment by hainm** [comment]
> dlopen really turns it into a runtime dependency, not a compile-time.

"Please install libsander to use this function"

vs

(current)
"Please reinstall cpptraj and install libsander to use this function"

**Comment by drroe** [comment]
Is `dlopen` portable?

**Comment by swails** [comment]
Should be.  OpenMM uses it and it's only disabled on the PNACL hardware (chromeos, I think).

Amber also uses it for the PLUMED plugin, which I've gotten to compile on Windows, Mac, and Linux.  So I'd say it's portable.

**Comment by drroe** [comment]
OK, a few simple tests has made me realize that learning how to use `dlopen` correctly is going to take me a bit, and starting 4 days before the code deadline is a bad idea. At this point I would be OK with having this functionality turned off in AmberTools by default. If people want it they can use cpptraj's standalone configure/make. I can even make it so that the standalone configure will use this automatically if a valid libsander.so is found. I just don't think there is enough time to thoroughly test this with all of the possible ways Amber can be built and on different platforms (would be nice if we at least still had Cruise Control builds...).

**Comment by hainm** [comment]
> starting 4 days before the code deadline is a bad idea. 

absolutely agree.

>  At this point I would be OK with having this functionality turned off in AmberTools by default.

agree.

**Comment by swails** [comment]
> At this point I would be OK with having this functionality turned off in AmberTools by default

For serial builds you'll be assured that libsander will be available, so why not turn it on at least for that?  And for parallel builds have configure2 look for libsander.so and turn it on if present (with a nice error message if not).  Otherwise you're looking at a year+ before people will actually start making use of this, and I think it's pretty useful.

**Comment by drroe** [comment]
> For serial builds you'll be assured that libsander will be available, so why not turn it on at least for that? And for parallel builds have configure2 look for libsander.so and turn it on if present

:+1: Both great suggestions. I'll work on implementing them.

**Comment by drroe** [comment]
Implemented in #257. If all goes well I will merge into Amber along with the additional changes needed to make it work.

**Comment by swails** [comment]
:+1:

**Comment by drroe** [comment]
#258 will ensure that libcpptraj.so does not include libsander. Hopefully this is the last tweak.

**Comment by hainm** [comment]
Can u merge to amber repo right after this PR so I can test? Thx.

**Comment by drroe** [comment]
@hainm This has now been pushed to the Amber GIT tree.

**Comment by hainm** [comment]
thx

**Comment by drroe** [comment]
This now seems to be working well. If any new problems crop up they can be addressed in a separate issue.

**Comment by hainm** [comment]
@drroe Action_Esander is still there. This is your intention right?

```bash
nm libcpptraj.so |grep sander
000000000011a300 T _ZN14Action_Esander4InitER7ArgListR10ActionIniti
000000000011a2e0 T _ZN14Action_Esander5PrintEv
000000000011a2a0 T _ZN14Action_Esander5SetupER11ActionSetup
000000000011a2c0 T _ZN14Action_Esander8DoActionEiR11ActionFrame
000000000011a280 T _ZN14Action_EsanderC1Ev
000000000011a280 T _ZN14Action_EsanderC2Ev
000000000011a3a0 W _ZN14Action_EsanderD0Ev
000000000011a340 W _ZN14Action_EsanderD1Ev
000000000011a340 W _ZN14Action_EsanderD2Ev
000000000011a320 T _ZNK14Action_Esander4HelpEv
000000000011a360 W _ZNK14Action_Esander5AllocEv
00000000006a47f0 V _ZTI14Action_Esander
00000000003c0da0 V _ZTS14Action_Esander
00000000006a47a0 V _ZTV14Action_Esander
```

**Comment by drroe** [comment]
Yes - but if you try to use it when `USE_SANDERLIB` is not defined it gives an error.

**Comment by hainm** [comment]
ok. just want to make sure.
